### PR TITLE
Create the temp dataset in the same location as the queried table in BigQueryTableRowIterator

### PR DESCRIPTION
**Pull request by peihe**
This fixed https://github.com/GoogleCloudPlatform/DataflowJavaSDK/issues/405

### CODE REVIEW

**Comment by peihe** [comment]
@dhalperi

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java [hunk 1]
```diff
@@ -435,11 +463,11 @@ private TableReference executeQueryAndWaitForCompletion()
         result = client.execute();
         break;
       } catch (IOException e) {
-        LOG.error(String.format(error, errorArgs), e.getMessage());
+        String errorMessage = error + String.format(" Cause: %s", e.getMessage());
+        LOG.error(errorMessage);
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java]: why is this not just `LOG.error("{}", error, e);`

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java [hunk 2]
```diff
@@ -435,11 +463,11 @@ private TableReference executeQueryAndWaitForCompletion()
...
         if (!BackOffUtils.next(sleeper, backOff)) {
-          LOG.error(
-              String.format(error, errorArgs), "Failing after retrying " + MAX_RETRIES + " times.");
-          throw e;
+          LOG.error(error + String.format(" Failing after retrying %d times.", MAX_RETRIES));
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java]: and similarly something like: `LOG.error("{}. Failing to execute job after {} attempts.", error, MAX_RETRIES+1)`

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java [hunk 3]
```diff
@@ -423,8 +451,8 @@ private TableReference executeQueryAndWaitForCompletion()
   // client - BQ request to be executed
   // error - Formatted message to log if when a request fails. Takes exception message as a
   // formatter parameter.
-  public static <T> T executeWithBackOff(AbstractGoogleClientRequest<T> client, String error,
-      Object... errorArgs) throws IOException, InterruptedException {
+  public static <T> T executeWithBackOff(AbstractGoogleClientRequest<T> client, String error)
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java]: this looks backwards-incompatible. Add a handler for the old method that indirects through `String.format`?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java (hunk 1 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java]: done
I want to build the error message once and use in logging and exception.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java (hunk 3 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java]: done

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java (hunk 2 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java]: done, but constructs with String.format in order to use it in both logging and exception.

**Comment by peihe** [comment]
PTAL

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java [hunk 4]
```diff
@@ -435,11 +474,12 @@ private TableReference executeQueryAndWaitForCompletion()
...
       } catch (IOException e) {
-        LOG.error(String.format(error, errorArgs), e.getMessage());
+        LOG.error("{}", error, e);
         if (!BackOffUtils.next(sleeper, backOff)) {
-          LOG.error(
-              String.format(error, errorArgs), "Failing after retrying " + MAX_RETRIES + " times.");
-          throw e;
+          String errorMessage = String.format(
+              "%s. Failing to execute job after %d attempts.", error, MAX_RETRIES + 1);
+          LOG.error(errorMessage);
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java]: `{}, errorMessage, e` ?

**Comment by dhalperi** [comment]
```
Results :
Failed tests: 
com.google.cloud.dataflow.sdk.util.BigQueryTableRowIteratorTest.testQueryFailed(com.google.cloud.dataflow.sdk.util.BigQueryTableRowIteratorTest)
  Run 1: BigQueryTableRowIteratorTest.testQueryFailed:241 
Expected: a string containing "bad query"
     but: was "Error when trying to dry run query NOT A QUERY.. Failing to execute job after 4 attempts."
  Run 2: BigQueryTableRowIteratorTest.tearDown:107 
No interactions wanted here:
-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIteratorTest.tearDown(BigQueryTableRowIteratorTest.java:107)
But found this interaction on mock 'mockClient':
-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.executeQueryAndWaitForCompletion(BigQueryTableRowIterator.java:391)
Actually, above is the only interaction with this mock.
```

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java (hunk 4 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/BigQueryTableRowIterator.java]: done

**Pull request merged**